lib/SPIRVProducerPass.cpp

```diff
@@ -4329,15 +4329,48 @@ void SPIRVProducerPass::GenerateInstruction(Instruction &I) {
   case Instruction::FCmp: {
     CmpInst *CmpI = cast<CmpInst>(&I);
 
-    // Pointer equality is invalid.
+    // Pointer comparisons are only supported from SPIR-V 1.4 onwards.
     Type *ArgTy = CmpI->getOperand(0)->getType();
     if (isa<PointerType>(ArgTy)) {
-      CmpI->print(errs());
-      std::string name = I.getParent()->getParent()->getName().str();
-      errs()
-          << "\nPointer equality test is not supported by SPIR-V for Vulkan, "
-          << "in function " << name << "\n";
-      llvm_unreachable("Pointer equality check is invalid");
+      if (SpvVersion() >= SPIRVVersion::SPIRV_1_4) {
+        SPIRVID cmp_lhs = getSPIRVValue(CmpI->getOperand(0));
+        SPIRVID cmp_rhs = getSPIRVValue(CmpI->getOperand(1));
+        spv::Op Opcode;
+        SPIRVOperandVec Ops;
+        switch (CmpI->getPredicate()) {
+        case CmpInst::ICMP_NE:
+          Opcode = spv::OpPtrNotEqual;
+          break;
+        case CmpInst::ICMP_EQ:
+          Opcode = spv::OpPtrEqual;
+          break;
+        case CmpInst::ICMP_UGT:
+        case CmpInst::ICMP_UGE:
+        case CmpInst::ICMP_ULT:
+        case CmpInst::ICMP_ULE:
+          Opcode = GetSPIRVCmpOpcode(CmpI);
+          Ops << getSPIRVType(Type::getInt32Ty(Context)) << cmp_lhs << cmp_rhs;
+          RID = addSPIRVInst(spv::OpPtrDiff, Ops);
```

Comment:
Something seems off here. OpPtrDiff returns a signed result.

a = 0xa;
b = 0xb

a < b should be true

diff = a - b = 0xffff (-1)
diff < 0 is false when the comparison is unsigned.

Should the comparison be converted to a signed opcode?

Reply:
You are right, thanks for catching this! I've added another utility function to map pointer comparison predicates to SPIR-V opcodes.